### ISSUE TITLE
Custom split normals no longer working

### ISSUE THREAD

**Issue by SteffenDuenner** [LuxCoreRender/BlendLuxCore#451]
Sorry to reopen an old issue ( https://github.com/LuxCoreRender/BlendLuxCore/issues/120 ) but the feature from Mar 30 is no longer working. I know it worked some weeks ago.

.blend file: 

[Custom_Split_Normals_v01.blend.zip](https://github.com/LuxCoreRender/BlendLuxCore/files/4618940/Custom_Split_Normals_v01.blend.zip)

Steps to reproduce:

- Open the attached blend file
- Select either the cube or the hood
- Go to Mesh > Normals and toggle the Auto Smooth checkbox
- See the difference it makes in the viewport using MatCaps (same for Cycles / Eevee):

![CustomSplitNormals](https://user-images.githubusercontent.com/33224667/81755343-cf3d0f00-94b8-11ea-81f7-dd82defb6443.gif)

- Now start a preview render and toggle the checkbox. The result is always this:

![Screenshot from 2020-05-13 01-15-16](https://user-images.githubusercontent.com/33224667/81755621-95b8d380-94b9-11ea-8c2e-eb771126af21.png)

- F12 rendering yields the same result. The cube's normals are controlled by a "Weighted Normal" modifier, the hood object was exported from CAD with vertex normals. Both completely ignore the Auto Smooth setting (which means they also ignore their custom split normals / vertex normals)

P.S.: Yes, the lighting is totally different because in the viewport I used a MatCap and with LuxCore I use 2 area lights because I didn't want to bundle HDRs with the blend file and I wanted to keep it as simple as possible. Main problem is that BlendLuxCore doesn't react to custom split normals.

OS: Linux Mint 19.3
Blender version: 2.90.1, very latest master
LuxCore version: latest Azure build ( https://github.com/LuxCoreRender/LuxCore/commit/08c9ade29fa84a9c9fe04df89fcf5db54b026b4d )

**Comment by Theverat** [comment]
Blender 2.90 will be supported once it is stable.
We always only support the latest stable Blender release.
Please reopen this issue if the issue persists once the first finished 2.90 release is out.

**Comment by SteffenDuenner** [comment]
OK, sorry. I just gave it a try with official 2.82a release and everything works as expected. Shame on me.

**Comment by Theverat** [comment]
No worries :)

**Comment by Theverat** [comment]
Reopening because it also happens in Blender 2.83, which is now the latest stable release.

The cause could be a change in the Blender mesh datastructure layout.
Probably requires changes in our mesh conversion C++ code.

**Comment by Theverat** [comment]
I should have fixed this with https://github.com/LuxCoreRender/LuxCore/commit/989ddc07f6900bf4054f12c0527701b1d734187f
Note that this breaks support for Blender 2.82 and older.

**Comment by SteffenDuenner** [comment]
Confirmed working! Thanks a lot for the fix!

**Comment by SteffenDuenner** [comment]
Confirmed working in 2.83! Thanks

In 2.90 master it's still not working. Looks like there will have to be another fix ;)

**Comment by Theverat** [comment]
BlendLuxCore is now only registered if the Blender version is supported. Currently, this is only 2.83.
Neither lower nor heigher versions are supported.

It makes no sense for us to suppport Blender 2.90 because it is in development, and their data types might change every day, which would require us to follow suit every time.

**Comment by SteffenDuenner** [comment]
Sure, I totally get this, no worries. I was just wondering what they do in every new version of Blender to break the normals again.

**Comment by Theverat** [comment]
Well, the Blender developers are not at fault here.
I am using interal data structures, which are under no obligation to stay the same, to get fast C++-accelerated custom normal export.
Unforutnately there is no stable public C API for Blender, which would be the best solution for this kind of problem.

**Comment by SteffenDuenner** [comment]
OIC! Thanks for clearing that up.

**Comment by Dade916** [comment]
@Theverat, can't we just keep a copy of the old function and conditional call, from python, the old or the new one according the Blender version ? It should require just couple of lines of code.

**Comment by Theverat** [comment]
I could do that.
This would mean that custom normals are fast only in the supported Blender version, and slow in other versions.
But even aside from custom normals, I would find it good to abort the addon registration in unsupported Blender versions, because there can be endless other issues with them, and people have in the past repeatedly opened bug reports about those which creates work for me (I need to investigate and eventually close them).

**Comment by Dade916** [comment]
> This would mean that custom normals are fast only in the supported Blender version, and slow in other versions.

Why ? You keep the old v2.82 code path and you will add a v2.83 new code path. It will work as before with v2.82 and the new code path will be used with v2.83 or better.

For what I understand, custom normals don't work at all with not supported Blender versions (it is not about being fast/slow).

P.S. Blender v2.83 doesn't work with the OIDN denoiser and Linux at the moment.

**Comment by Theverat** [comment]
> Why ? You keep the old v2.82 code path and you will add a v2.83 new code path. It will work as before with v2.82 and the new code path will be used with v2.83 or better.

The problem is that some Blender structs were changed between versions.
I guess I could create separate struct declarations for each Blender version we support and use the appropriate one at runtime. It would make the code more complicated though and would create LOTS of duplicate code (around 350 lines per Blender version, everything below here would need to be duplicated: https://github.com/LuxCoreRender/LuxCore/blob/6f9716533448f70096355131946eb702a4a20784/include/luxcore/pyluxcore/blender_types.h#L98

> For what I understand, custom normals don't work at all with not supported Blender versions (it is not about being fast/slow).

No, it was simply coincidence that the function that finds the custom normals layer just returned -1. In Blender 2.82, the new code crashes.

**Comment by Theverat** [comment]
I have now added custom normals support for Blender 2.82.7 (aka 2.82a) and 2.82.0.
It will be possible to support more versions in the future.

**Comment by SteffenDuenner** [comment]
Now that Blender 2.91 has been released some weeks ago and 2.92 is about to be released soon, may I bump this thread to get working custom split normals working again with BlendLuxCore?

**Comment by juangea** [comment]
Yep, it could be great for 2.92 and 2.5 :)

**Comment by Theverat** [comment]
I have added a slower Python fallback in 7136652e802cba11832f8196d0008592d3a56d0a and https://github.com/LuxCoreRender/LuxCore/commit/f2642e1164b36e281a2f0367eb2d4708dfb98f56 so even if there exists no fast C++ path for a particular Blender version, custom normals will continue to work (just slower for heavy meshes). A warning is shown if this fallback causes a mesh to export more than 0.3 seconds slower.

The main reason is that the Blender mesh has proven to be a much faster moving target than I anticipated, almost every new Blender version would require us to add C++ code to support it (which is not super trivial). So I guess we will only add C++ support for LTS versions of Blender, or something like that.
